Extracted Logic in Pairing Table View to Presenter

yusufjimoh opened the pull request:
### Description of work
- extracted business logic from the pairing table view to the presenter.
- cleaned up some code in the view and presenter.
- updated unit tests to reflect changes in the view and presenter

Fixes #38105 

### To test:
- testing instructions are included in the  associated issue
- code should pass unit tests

This does not require release notes because it involves code restructuring, no new functionality has been added

yusufjimoh committed:
Extracted Logic in Pairing Table View to Presenter

yusufjimoh committed:
Updated Failing Tests

- Updated Failing Tests
- Created Getters for Properties in PairingTableView so the properties are accesed through the getters in the PairingTablePresenter

adriazalvarez commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_presenter.py:
```diff
@@ -220,28 +262,173 @@ def remove_selected_rows_in_view_and_model(self, pair_names):
+        guess_alpha_widget.clicked.connect(self._view.guess_alpha_clicked_from_row)
+        self._view.set_widget_in_table(row, 5, guess_alpha_widget)
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # Context Menu
+    # ------------------------------------------------------------------------------------------------------------------
+    def _context_menu_event(self, _event):
+        """Overridden method for dealing with the right-click context menu"""
+        menu = self._view.create_context_menu
+
+        add_pair_action = self._context_menu_add_pair_action(self._view.add_pair_button.clicked.emit)
+        remove_pair_action = self._context_menu_remove_pair_action(self._view.remove_pair_button.clicked.emit)
+
+        if self._view.is_disabled:
+            self._view.add_pair_action.setEnabled(False)
+            self._view.remove_pair_action.setEnabled(False)
+        # set-up the menu
+        menu.addAction(add_pair_action)
+        menu.addAction(remove_pair_action)
+        menu.popup(self._view.cursor_position)
+
+    def _context_menu_add_pair_action(self, slot):
+        add_pair_action = self._view.get_add_pair_action
+        add_pair_action.setCheckable(False)
+        if len(self.get_selected_row_indices()) > 0:
+            add_pair_action.setEnabled(False)
+        add_pair_action.triggered.connect(slot)
+        return add_pair_action
+
+    def _context_menu_remove_pair_action(self, slot):
+        if len(self.get_selected_row_indices()) > 1:
+            # use plural if >1 item selected
+            remove_pair_action = self._view.get_remove_pair_action(True)
+        else:
+            remove_pair_action = self._view.get_remove_pair_action()
+        if self._view.get_pairing_table.rowCount() == 0:
+            remove_pair_action.setEnabled(False)
+        remove_pair_action.triggered.connect(slot)
+        return remove_pair_action
+
```
These functions, even though they contain a bit of logic, they are creating q elements and connecting slots/signals, perhaps is better to move them to the view.

adriazalvarez commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_presenter.py:
```diff
@@ -13,8 +13,8 @@
+from mantidqtinterfaces.Muon.GUI.Common.utilities import table_utils
 
 
 class PairingTablePresenter(object):
```
This class is connecting slots/signals when initialized. 
On the spirit of the MVP training, could you create a subscriber function on the view, and subscribe the presenter to it? Also, create notifier functions on the view that will call the handler's in the presenter.

adriazalvarez commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_constants.py:
```diff
@@ -0,0 +1,18 @@
+inverse_pair_columns = {"pair_name": 0, "to_analyse": 1, "group_1": 2, "group_2": 3, "alpha": 4, "guess_alpha": 5}
+
+
+def get_index_of_text(selector, text):
```
I have found this function declared exactly the same in other parts of MUON interfaces. This is outside the scope of this issue, but can we create an issue to investigate wheter we can declare this function on some utils file of the utilities folder of the class?

adriazalvarez replied:
Added #38315

adriazalvarez reviewed (changes requested):
Thanks for the effort on moving the logic out of the view class. I know this view is not easy to refactor !  Generally the view seems more neat and decluttered, although I have found some moved functions that maybe belong in the view, and we could subscribe the presenter to the view to avoid connecting signals/slots on the presenter and use notifiers instead to call the relevant functions on the presenter. 

Also, when testing I have found this error when i click on `Load current run` : 

![image](https://github.com/user-attachments/assets/19d12dc4-6300-4fe5-bb68-d5823b664938)

yusufjimoh committed:
Add Unit Test for new methods in presenter

yusufjimoh committed:
--amend-no-edit

adriazalvarez commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_constants.py:
I think we don't need a new file for storing two dictionaries. 
What I would do is to keep the definition of the pair_columns and inverse pair_colums on the view, as well as the function get_pair_columns. 
Then import the dictionary and the function to the presenter as you would do from the constant file. 
`from mantidqtinterfaces.Muon.GUI.Common.pairing_table_widget.pairing_table_widget_view import get_pair_columns,pair_columns`

Regarding the get_index_of_Text function, I'll add a comment on the test (where you use it)

adriazalvarez commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_presenter.py:
```diff
@@ -13,22 +13,16 @@
-        self._view.on_guess_alpha_clicked(self.handle_guess_alpha_clicked)
-        self._view.on_table_data_changed(self.handle_data_change)
+        self._view.subscribe(self)
+        self._view.notify()
```
Thanks for adding this two function, it looks good. 
Perhaps we could rename the function `notify` to be `subscribe_notifiers_to_presenter` or something of the sort.

adriazalvarez commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_presenter.py:
```diff
@@ -13,22 +13,16 @@
 
 
 class PairingTablePresenter(object):
     def __init__(self, view, model):
```
On the constructor for this class, I think these two slots are not used, could you remove them? 
```
        self._on_alpha_clicked = lambda: 0
        self._on_guess_alpha_requested = lambda pair_name, group1, group2: 0
```

yusufjimoh replied:
I have removed the code from the constructor

adriazalvarez commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_presenter.py:
```diff
@@ -49,39 +43,81 @@ def notify_data_changed(self):
             if not self.validate_alpha(changed_item_text):
                 update_model = False
             else:
                 rounded_item = round_value(changed_item_text, self._model._context.group_pair_context.alpha_precision)
```
In the presenter class, a private member of the model : `self.model._context` is accesed in a few places. Can you add the necessary getter functions on this model so that the context is not accesed directly?

adriazalvarez commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_view.py:
```diff
@@ -203,32 +252,36 @@ def _guess_alpha_button(self):
+
+    def set_item_selectable_and_enabled(self, row, col):
+        item = self.get_table_item(row, col)
+        item.setFlags(self.get_default_item_flags())
```
This line is giving an error on runtime as you are trying to access a property..
```suggestion
        item.setFlags(self.get_default_item_flags)
```

adriazalvarez commented on qt/python/mantidqtinterfaces/test/Muon/grouping_tab/pairing_table_group_selector_test.py:
```diff
@@ -15,6 +15,7 @@
 from mantidqtinterfaces.Muon.GUI.Common.pairing_table_widget.pairing_table_widget_presenter import PairingTablePresenter
 from mantidqtinterfaces.Muon.GUI.Common.pairing_table_widget.pairing_table_widget_view import PairingTableView
 from mantidqtinterfaces.Muon.GUI.Common.test_helpers.context_setup import setup_context_for_tests
+from mantidqtinterfaces.Muon.GUI.Common.pairing_table_widget.pairing_table_constants import get_index_of_text
```
Although I have opened an issue to investigate the reuse of get_index_of_text accross muon (#38315 ). The `get_index_of_text` function hsa been declared as a static method on the elemental analysis table view class, so you can access directly without creating the table object. 
```suggestion
from mantidqtinterfaces.Muon.GUI.ElementalAnalysis2.grouping_widget.ea_grouping_table_widget_view import EAGroupingTableView
```

adriazalvarez commented on qt/python/mantidqtinterfaces/test/Muon/grouping_tab/pairing_table_group_selector_test.py:
```diff
@@ -93,14 +94,14 @@ def test_that_added_groups_appear_in_group_combo_boxes_in_new_pairs(self):
         self.presenter.handle_add_pair_button_clicked()
 
-        index = self.view.get_index_of_text(self.get_group_1_selector_from_pair(0), "random string")
+        index = get_index_of_text(self.get_group_1_selector_from_pair(0), "random string")
```
```suggestion
        index = EAGroupingTableView.get_index_of_text(self.get_group_1_selector_from_pair(0), "random string")
```

yusufjimoh replied:
suggested changes has been applied, thanks

adriazalvarez reviewed (changes requested):
Except for a typo that is creating a runtime error, I have followed the manual testing instructions without any functional issue on the pair table widget. Thanks for the good work on this refactoring. 
I have added some additional comments on small things that I've noticed during review.

yusufjimoh committed:
Apply Suggested Changes from code review to PairingTableWidget View and Presenter
- created a getter for model._context so the property can be accessed using the getter
- remove pairing table constants

yusufjimoh committed:
renamed overidden context_menu_event back to camelcase

yusufjimoh commented:
Code has been updated to fix an issue on the pairing table pointed out by @adriazalvarez, Thanks.

adriazalvarez reviewed (changes requested):
Thanks for the latest changes Yusuf. 
I have been testing the pairing table again, everything works as expected.

adriazalvarez reviewed (dismissed):
I forgot to approve before!

cailafinn commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_view.py:
```diff
@@ -116,84 +126,128 @@ def set_up_table(self):
+    def set_widget_in_table(self, row, col, widget):
+        self.pairing_table.setCellWidget(row, col, widget)
+
+    def set_checkbox_in_table(self, item, checked):
+        if checked:
+            item.setCheckState(QtCore.Qt.Checked)
+        else:
+            item.setCheckState(QtCore.Qt.Unchecked)
+
```
This seems like it could use `setChecked` and use the `checked` boolean directly.

cailafinn commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_view.py:
```diff
@@ -268,126 +326,80 @@ def on_cell_changed(self, row, column):
+        """
+        Create and return the 'Remove Pair' or 'Remove Pairs' QAction based on the plural flag.
+        """
+        if plural:
+            self.remove_pair_action = QtWidgets.QAction("Remove Pairs", self)
+            return self.remove_pair_action
+        else:
+            self.remove_pair_action = QtWidgets.QAction("Remove Pair", self)
+            return self.remove_pair_action
```
This looks like it just could be 
```suggestion
        self.remove_pair_action = QtWidgets.QAction("Remove Pairs", self) if plural else QtWidgets.QAction("Remove Pair", self)
        return self.remove_pair_action
```

cailafinn reviewed (changes requested):
Just noticed a couple more improvements that could be made while we're editing these files.

yusufjimoh commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_presenter.py:
```diff
@@ -49,39 +40,81 @@ def notify_data_changed(self):
+        pair_name = self.get_table_item_text(row, 0)
         update_model = True
-        if pair_columns[col] == "pair_name" and not self.validate_pair_name(changed_item_text):
+        if get_pair_columns()[col] == "pair_name" and not self.validate_pair_name(changed_item_text):
             update_model = False
```
I was thinking that the return statement might not be ideal because line 127, which is the final condition that checks if col != 1, would not be triggered if a return statement is used. What are your thoughts?

if col != 1:
            self.update_view_from_model()
            self.notify_data_changed()

cailafinn replied:
Ah, good catch, that part was hidden by github so I hadn't noticed! Thanks

yusufjimoh commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_presenter.py:
```diff
@@ -220,28 +253,102 @@ def remove_selected_rows_in_view_and_model(self, pair_names):
+    # ------------------------------------------------------------------------------------------------------------------
+    # Enabling / Disabling the table
+    # ------------------------------------------------------------------------------------------------------------------
+    def _disable_all_table_items(self):
+        for row in range(self.num_rows()):
+            for col in range(self.num_cols()):
+                column_name = pair_columns[col]
+                if column_name in ["group_1", "group_2", "guess_alpha"]:
+                    self._view.set_widget_enabled(row, col, False)
+                else:
+                    self._view.set_item_selectable(row, col)
+
+    def _enable_all_table_items(self):
+        for row in range(self.num_rows()):
+            for col in range(self.num_cols()):
+                column_name = pair_columns[col]
+                if column_name in ["group_1", "group_2", "guess_alpha"]:
+                    self._view.set_widget_enabled(row, col, True)
+                elif column_name == "pair_name":
+                    self._view.set_item_selectable_and_enabled(row, col)
+                elif column_name == "alpha":
+                    self._view.set_item_editable_and_enabled(row, col)
+                elif column_name == "to_analyse":
+                    self._view.set_item_checkable_and_enabled(row, col)
```
Yes, the first method enables the widgets while the second disables them but I have refactored the code to address the duplicated logic.

yusufjimoh commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_presenter.py:
```diff
@@ -220,28 +253,102 @@ def remove_selected_rows_in_view_and_model(self, pair_names):
+            item = self._view.create_table_item(entry, color, tooltip)
+            column_name = pair_columns[i]
+
+            if column_name == "pair_name":
+                self._view.add_pair_name_entry(row_position, i, entry)
+            if column_name == "group_1":
+                self._view.add_group_selector_entry(row_position, i, entry, group="group_1")
+            if column_name == "group_2":
+                self._view.add_group_selector_entry(row_position, i, entry, group="group_2")
+            elif column_name == "alpha":
+                self._view.add_alpha_entry(row_position, i, entry)
+            if column_name == "to_analyse":
+                self._view.add_to_analyse_checkbox(item, entry)
```
the elif statement has been changed to if statement

yusufjimoh commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Muon/GUI/Common/pairing_table_widget/pairing_table_widget_presenter.py:
```diff
@@ -49,39 +40,81 @@ def notify_data_changed(self):
+
+    def enable_updates(self):
+        """Allow update signals to be sent."""
+        self._view.set_is_updating(False)
```
@adriazalvarez what do you think about this, do you think the function should be renamed?

adriazalvarez replied:
Yes, probably the naming is confusing in this case, although I see how it is name as it is, as the function really means enable signals, and `is_updating` is a flag to not process view signals. Perhaps a clearer name could be `unblock_update_signals` and `block_update_signals` for `disable_updates`. 
What do you think, @cailafinn ?

cailafinn replied:
I think the best way would be to rename the functions in the view to be `disable_updates` and `enable_updates`. `is_updating` and the term `signals` are implementation details of the view, so ideally should not be used as terms in the view's public-facing API.

yusufjimoh committed:
Improve code using features in python 3.10

cailafinn reviewed (changes requested):
Apologies for the late review, I didn't see the notification that the changes had been made. 

This implementation of the context manager won't work, unfortunately. Where the `with` blocks are isn't where the updates need to be stopped. Instead, they should be used in places like `add_pair_to_view`
```python
    def add_pair_to_view(self, pair, to_analyse=False, color=row_colors[RowValid.valid_for_all_runs], tool_tip=row_tooltips[RowValid.valid_for_all_runs]):
        with self.disable_updates_context():
            self.update_group_selections()
            assert isinstance(pair, MuonPair)
            entry = [str(pair.name), to_analyse, str(pair.forward_group), str(pair.backward_group), str(pair.alpha)]
            self.add_entry_to_table(entry, color, tool_tip)
```
Which would disable the updates to the view during the time the execution is inside the `with` block. `update_view_from_model` would be the other place you'd want to use the context manager. Let me know if you have any questions.

yusufjimoh committed:
modified some presenter functions to use context manager

renamed enable updates method to disable updates method and vice versa

cailafinn reviewed (approved):
Double-checked functionality and seems to be working well, nice work.
